### ISSUE TITLE
Using docker from inside the default Codespaces image fails with a permission error or a notice docker is not running

### ISSUE THREAD

**Issue by Chuxel** [microsoft/vscode-dev-containers#727]
The [codespaces-linux](https://github.com/microsoft/vscode-dev-containers/tree/master/containers/codespaces-linux) image switched from using the [docker-from-docker/docker-outside-of-docker](https://github.com/microsoft/vscode-dev-containers/tree/master/containers/docker-from-docker) pattern to doing true [docker-in-docker](https://github.com/microsoft/vscode-dev-containers/tree/master/containers/docker-in-docker) as of version 1.1.0 of the image.  This model has advantages particularly in the Codespaces environment because you can "bind" mount from anywhere in the codespace into your containers - which some CLIs and extensioons require.

Unfortunately, this turned out to be a **breaking change** for the definition since the `--priviledged` flag needs to be included in runArgs in devcontainer.json. (As reported [here](https://github.com/microsoft/vscode-dev-containers/issues/717) and [here](https://blog.jongallant.com/2021/01/vscode-codespaces-container-is-not-running-permission-denied/).)

To resolve this problem, update your devcontainer.json from:

```json
"runArgs": [ "--cap-add=SYS_PTRACE", "--security-opt", "seccomp=unconfined"],
```
...to...
```json
"runArgs": [ "--cap-add=SYS_PTRACE", "--security-opt", "seccomp=unconfined", "--privileged", "--init" ],
```

You can also remove the `mount` property as it is no longer needed, but leaving it there does not cause harm. We'll be sure to bump the major image version number in these situations in the future.

**Comment by erinxocon** [comment]
@Chuxel I think this might be something I am experiencing. I placed this issue https://github.com/microsoft/vscode-dev-containers/issues/726. When I built the docker-in-docker on my linux machine locally, I was able to run docker...however things got hairy when I tried to follow some of the documentation/scripts inside the container. Let me know if you would like for me to try something out locally!

**Comment by Chuxel** [comment]
@erinxocon Yeah, saw that issue as well - @2percentsilk was planning to respond there, but yes, the README is wrong. That code needs to be in an ENTRYPOINT script to work which is what the definition actually does.

To clarify this issue, however, there's now two separate definitions:

* docker-from-docker - This uses your local docker engine from inside the container. 
* docker-in-docker - This is a brand new definition that should be setup correctly that actually runs the docker engine inside the container itself. This adds overhead, but better bind mounting support. It requires `--priviledged` in `runArgs` to function.

This particular issue relates to the codespaces-linux definition, which did not previously include the `--priviledged` flag and now uses the approach in the new docker-in-docker definition.

**Comment by erinxocon** [comment]
@Chuxel Got it, that makes sense. I was saw the new docker-in-docker when I cloned the repository but didn't bother with it as I just wanted the docker-from-docker. The terminology is a bit confusing to me as I use the docker:dind image in CI all the time and that is what y'all're calling docker-from-docker. Thanks for the clarification!

**Comment by Chuxel** [comment]
@erinxocon Actually - you nailed it on the head. docker-from-docker isn't docker:dind.   docker-from-docker is forwarding the local docker socket and only the CLI is in the container. docker:dind has the docker engine inside, which is what the new docker-in-docker definition does. Bind-mounting is the one downside of docker-from-docker, but DinD has a number potential issues depending on your setup. See [here](https://jpetazzo.github.io/2015/09/03/do-not-use-docker-in-docker-for-ci/). But for the codespaces image, DinD is the better choice.

**Comment by erinxocon** [comment]
@Chuxel Gosh! I was confusing my terminology again! When I setup the CI/CD for my company we _were_ originally using dind, we now pass the context into the containers a la docker-from-docker and socket binding! We were using the docker image to avoid having to install all the docker tools, however I've rolled our own base image. Long day :sweat_smile: Insert Xzibit Yo Dawg docker meme here.
